[iamtr] iP

iamtr opened the pull request:
# **Joseph Joshtur!!!**

> "Yare yare daze..." - Joseph Joestar

JJ frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~fast~~ _SUPER_ FAST to use

All you need to do is, 

1. download it from [here](https://www.google.com/)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

***

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

iamtr committed:
Changed greeting message for Duke

Changed name to Joseph and also added greeting messages for Level 0

iamtr committed:
Completed Level 1 - Echo

Program now will echo whatever user types

iamtr committed:
Created an todolist ArrayList and addItem method

iamtr committed:
Implemented handleInput() method

iamtr committed:
Handled issue where item is added even though there is no text

iamtr committed:
Created Task.java for storing and managing task data

iamtr committed:
Removed extra newline where calling 'list' command

iamtr committed:
Added getTaskDescription method for Task.java

iamtr committed:
Changed todolist from type string to a Task type

- Changed relevant helper methods as well

iamtr committed:
Add markAsDone() method

This changes the isDone boolean in Task to true

iamtr committed:
Refactor variable name "command" to "input"

iamtr committed:
Modify markAsDone() method to accept an argument

iamtr committed:
Add mark and unmark functionality

iamtr committed:
Rename getTaskDescription() to override toString()

iamtr committed:
Added Todo.java

iamtr committed:
Change toString() implementation for Todo.java

iamtr committed:
Create Deadline.java

iamtr committed:
Create Event.java class

iamtr committed:
Remove unneccessary space for toString() methods

iamtr committed:
Add handling of todo, deadline and event commands

iamtr committed:
Fix output when todolist is empty

When todolist is empty, the case is now handled by outputting "Todolist is empty"

iamtr committed:
Add "bye" command

Check if command is bye, then exits program if command is bye

iamtr committed:
Add constructor to DukeException

iamtr committed:
Add exception handling using DukeException

iamtr committed:
Add case to handle delete command

iamtr committed:
Create duke.txt to store todolist data

iamtr committed:
Import file writing libraries

iamtr committed:
Create method getDataString() for all Tasks

iamtr committed:
Add method saveData()

saveData() saves the todolist in a text file located at ip/data/duke.txt

iamtr committed:
Add method saveData()

saveData() saves the todolist in a text file located at ip/data/duke.txt

iamtr committed:
Merge branch 'branch-Level-7' of https://github.com/iamtr/ip into branch-Level-7

iamtr committed:
Change naming of constant from sep to SEP

iamtr committed:
Add file error handling for saveData() method

If directory or duke.txt is not present, it will be created by the error handling code

iamtr committed:
Change isDone to show 0 or 1 instead of "true" or "false" in getDataString()

iamtr committed:
Add basic test cases for automating the testing of text UI

iamtr committed:
Change tasks to accept a date format for arguments

Example: deadline /by 2002-01-01. For invalid arguments, exception handling is being done

iamtr committed:
Handle bug where \by is not specified in Deadline class

iamtr committed:
Remove unused imports

iamtr committed:
Implement helper classes

Parser class: Interprets input
Storage class: Handles storage of files
Tasklist class: Handles todolist
Ui class: Handles input and output to terminal

iamtr committed:
Refactor Duke.main() to use created classes

iamtr committed:
Remove unnecessary imports and old code

iamtr committed:
Change Duke to use OOP classes created

iamtr committed:
Remove unnecessary DukeException

iamtr committed:
Add getDescription() to Task

iamtr committed:
Fix getDataString() output

Fixed getDataString() which output the correct string

iamtr committed:
Add boolean isDone to task constructors

iamtr committed:
Add Tasklist in Duke constructor

iamtr committed:
Add loadData() method

iamtr committed:
Add restoreData() method

iamtr committed:
Update Task constructors in Parser class

iamtr committed:
Create new packages for seperate functionalities

iamtr committed:
Merge remote-tracking branch 'origin/add-gradle-support'

iamtr committed:
Update build.gradle

iamtr committed:
Add checkstyle files

iamtr committed:
Add ParserTest for testing Parser

iamtr committed:
Add UiTest for testing UI

iamtr committed:
Fix checkstyle errors

iamtr committed:
Modify Javadoc for run()

iamtr committed:
Update run() JavaDoc

iamtr committed:
Merge branch 'branch-A-JavaDoc'

macareonie commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,54 @@
+ * Represents the storage for the tasklist.
+ */
+public class Storage {
+    private static final Path path = Paths.get("data", "duke.txt");
```
Constant names should be all uppercase using underscore to separate words
```suggestion
    private static final Path PATH = Paths.get("data", "duke.txt");
```

NeoHW replied:
private final constants should be uppercase
```suggestion
    private static final Path PATH = Paths.get("data", "duke.txt");
```

macareonie commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class Ui {
+    private static final String SEP = "\t__________________________________________";
+    private final Scanner in;
+    private final PrintStream out;
```
Constant names should be all uppercase using underscore to separate words
```suggestion
    private final Scanner IN;
    private final PrintStream OUT;
```

NeoHW replied:
private final Constants should be in uppercase, with these classes names as nouns

```suggestion
    private final Scanner SCANNER_IN;
    private final PrintStream PRINTSTREAM_OUT;
```

macareonie commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,58 @@
+    private Parser parser;
+    private Ui ui;
+    private Storage storage;
+    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
```
Constant names should be all uppercase using underscore to separate words
```suggestion
    private final ByteArrayOutputStream OUT_CONTENT = new ByteArrayOutputStream();
```

NeoHW replied:
private final constants should be uppercase, with separating underscores

```suggestion
    private final ByteArrayOutputStream OUT_CONTENT = new ByteArrayOutputStream();
```

macareonie commented on src/test/java/duke/ui/UiTest.java:
```diff
@@ -0,0 +1,92 @@
+ * Represents a test class for Ui
+ */
+public class UiTest {
+    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
+    private final PrintStream originalOut = System.out;
```
Constant names should be all uppercase using underscore to separate words
```suggestion
    private final ByteArrayOutputStream OUT_CONTENT = new ByteArrayOutputStream();
    private final PrintStream ORIGINAL_OUT = System.out;
```

NeoHW replied:
private final constants should be uppercase

```suggestion
    private final ByteArrayOutputStream OUT_CONTENT = new ByteArrayOutputStream();
    private final PrintStream ORIGINAL_OUT = System.out;
```

macareonie reviewed (commented):
Code mostly adheres to code quality guidelines with regards to naming. Only changes would be uppercasing the constant variables. Other than that, the classes, variables and methods are appropriately named. 👍

NeoHW commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,43 @@
+     * Runs the Duke application.
+     * Keeps receiving commands from the user and executing them until the user types "bye".
+     */
+    public void run() {
```
Good abstraction layer

NeoHW commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,125 @@
+            taskNumber = Integer.parseInt(details) - 1;
+            Task t = todolist.removeItem(taskNumber);
+            output = "Noted. I've removed this task:" + t.toString();
+            break;
```
you could consider abstracting out each case to make the code cleaner, and also for parts that require reuse

NeoHW commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,125 @@
+        String output = "";
+        int taskNumber;
+
+        switch (command) {
+        case "list":
```
i like that you remembered the switch case indentation form!

NeoHW commented on src/main/java/duke/task/Tasklist.java:
```diff
@@ -0,0 +1,109 @@
+
+    public boolean isEmpty() {
+        return todolist.isEmpty();
+    }
```
you could consider adding javadocs to these public methods too!

NeoHW commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @param message The message to be printed.
+     */
+    public void printMessage(String... message) {
```
I really like the usage of varags here!

NeoHW reviewed (approved):
Overall, I found your code easy to read. I noted a few minor coding standard violations
too, usually with regard to the naming of constants.
Some of the classes are getting quite long. You could consider splitting into
smaller classes and abstracting methods out if that makes sense.

iamtr committed:
Move parser to constructor instead of run()

iamtr committed:
Add Tasklist parameter for parseCommand()

parseCommand() now takes in a Tasklist parameter. Previously, Parser declares its own tasklist, which is not ideal as Parser should not have knowledge of tasklists, and its job is only to process commands

iamtr committed:
Refactor Task subclasses and Parser

If the description that a task is being given is empty, a DukeException will now be thrown

parseCommand() from Parser is now split into smaller methods for readability. These two changes were done simultaneously as there was no way to refactor them separately.

iamtr committed:
Refactor variable to smaller scope

iamtr committed:
Refactor restoreData() to smaller methods

iamtr committed:
Refactor parseCommand to return void

Checking if command is bye is now done by a seperate method commandIsBye()

iamtr committed:
Add method find

iamtr committed:
Add DukeException to findTask()

iamtr committed:
Add exception handling for handleDelete()

iamtr committed:
Add exception handling

iamtr committed:
Add dependencies for JavaFX on Gradle

iamtr committed:
Add .fxml files for GUI

iamtr committed:
Add images for dialog box

iamtr committed:
Add Main and MainWindow class

These class facilitates the

iamtr committed:
Add Launcher class

iamtr committed:
Change parseCommand to return String

I am aware that this is not an ideal way of doing things, since parseCommand is not supposed to interact with UI directly or return anything. However, given the scope of the project and the amount of time I want to spend here, I believe this amendment is fine for now. It is not good practice.

iamtr committed:
Add Javadocs for Launcher

iamtr committed:
Update README.md

iamtr committed:
Update README.md

iamtr committed:
Add assert statements

iamtr committed:
Add sort functionality

iamtr committed:
Merge pull request #2 from iamtr/branch-C-Sort

Add sort functionality

iamtr committed:
Change Duke to Joseph

iamtr committed:
Add User Guide

iamtr committed:
Modified error message

iamtr committed:
Create Ui.png

iamtr committed:
Add javadocs
